src/Symfony/Component/JsonEncoder/DependencyInjection/EncodablePass.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\JsonEncoder\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * Sets the encodable classes to the services that need them.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ */
+class EncodablePass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('json_encoder.encoder')) {
+            return;
+        }
+
+        $encodableClassNames = [];
+
+        // retrieve services tagged with "json_encoder.encodable" tag
+        // these are mostly loaded by "FrameworkExtension::registerJsonEncoderConfiguration()"
+
+        foreach (array_keys($container->findTaggedServiceIds('json_encoder.encodable')) as $id) {
+            if ($className = $container->getDefinition($id)->getClass()) {
+                $encodableClassNames[] = $className;
+            }
+
+            $container->removeDefinition($id);
+        }
+
+        // retrieve services within encodable namespaces
+        // these are loaded by user custom service definition
+
+        $encodableNamespaces = array_keys($container->getParameter('.json_encoder.encodable_paths'));
+
+        foreach ($container->getDefinitions() as $definition) {
```

Comment:
I don't understand why we need this style of registration. It looks like an anti-pattern to me.

Reply:
The use case here is: "I want to get classes located in namespaces".

To me, there are two main paths:
- the classes aren't services, so the `LoaderInterface::registerClasses()` + a tag, will work like a charm
- the classes already are services, so I need to retrieve them, and the only way I found is by iterating over services

I agree, that looks like an antipattern. But is it possible to retrieve already registered services that are located in a specific namespace in another way?

But maybe we can require the encodable classes to not be services? (or services with the proper tag)

Reply:
> The use case here is: "I want to get classes located in namespaces".

this can be achieved by loading the corresponding directory and tagging everything in it while doing so:
```
services:
  App\MyNamespace\:
    tags: ['json_encoder.encodable']
```

Reply:
Ok, so maybe we can remove the `framework.json_encoder.paths` config option, and instead ask the user to create such tagged services?

I could be great because in that way we can leverage service autoconfiguration and attribute autoconfiguration!

Reply:
yes, let's do this